### ISSUE TITLE
Running local using multiple threads (CPUs)?

### ISSUE THREAD

**Issue by ohdongha** [hillerlab/make_lastz_chains#2]
Hello, @kirilenkobm and @MichaelHiller, 
Thanks for creating this tool - I was looking for an option to run `lastz` genome-to-genome alignment on multiple threads and ended up here. 

I was submitting a job to a single SGE node that has 16 CPUs.  Since I am using only one machine and not going to `qsub` from the node, I used the default ( `--executor local` ) option (`--executor sge` switched back to local anyway when it realized `qsub` is not allowed in the node).  

Nextflow documentation says 
> The local executor is used by default. It runs the pipeline processes in the computer where Nextflow is launched. The processes are parallelised by spawning multiple threads and by taking advantage of multi-cores architecture provided by the CPU 

But looking at the temp. folders and files during the run, I am not sure whether all available threads are being used.  It seems like the query sequence is divided to 4 chunks (this might be decided by the chunk size settings rather than number of threads, I guess) and being aligned to the target.  But the 4 chunks appears to be aligned one after another, rather than in parallel.  

...

The following was the command line I used: 
``` 
make_chains.py Dme Dsi Dme_genomic.fna Dsi_genomic.fna --project_dir DmeDsi_1stAttempt
```
I also tried to convey that there are 16 CPUs, but this ran exactly the same way (four chunks of query being processed one after another):
```
make_chains.py Dme Dsi Dme_genomic.fna Dsi_genomic.fna --executor local --cluster_parameters "cpus=16" --project_dir DmeDsi_1stAttempt2
```

Is there a way to run this pipeline on multiple threads in parallel on a local machine?  Or is each of the four chunks somehow already being aligned using all available CPU threads? 

Thanks again!

**Comment by MichaelHiller** [comment]
Hi, 
if I understand it correctly, you want to run the alignment as a single job but multi-threaded to use your 16 CPUs. 
Lastz and the other steps do not support multi-threading, so every job would run on a single core. And if you don't use your node as a compute node that will handle several independent jobs in parallel, this means it will run sequentially.

My suggestion is to configure your SGE to use this node as a compute node with 16 cores. Alternatively, if this is a local server, just install Slurm instead of SGE. 

A note: For a mammal to mammal alignment, we need ~1000 CPU hours in total. Depending on what you want to align with 16 cores, this may take a while. 

Michael

**Comment by kirilenkobm** [comment]
Hi!

Thanks for your inquiry.
I need to do some testing to check whether it's possible in the current make_chains.py setup.
I will try to push the make_chains.py job to a single node and check whether nextflow can utilize the node's CPUs within the "local executor" framework. In the best case, Nextflow will "think" that it runs on a machine with 16 CPUs and execute lastz jobs smoothly.

I will let you know about the results later.

Bogdan

**Comment by ohdongha** [comment]
Thanks, @MichaelHiller and @kirilenkobm for the responses!

My understanding was that although LastZ cannot use more than a single core, your script would chunk the query sequence and create multiple jobs that align each chunk to the target sequence.  Then I hoped that the Nextflow could send each job to each core on a local machine (= a node in a SGE cluster), send another when one is finished, etc.  

> In the best case, Nextflow will "think" that it runs on a machine with 16 CPUs and execute lastz jobs smoothly.
So yes, this would be the ideal scenario :)  Please let me know how the test goes. 

> My suggestion is to configure your SGE to use this node as a compute node with 16 cores.
I am not very familiar with our SGE configuration of our cluster farm (and also first time with Nextflow).  I believe all nodes we submit our jobs should be already configured to use multiple cores (up to 64).  

Perhaps I should have run the `make_chains.py` script on the "head node" and let the nextflow to submit (qsub) to multiple computation nodes with 1-core (since LastZ cannot run on multiple)?   I will ask people who maintains our cluster. 

> A note: For a mammal to mammal alignment, we need ~1000 CPU hours in total. Depending on what you want to align with 16 cores, this may take a while.
Yes, we plan to compare mammalian genome assemblies.  If `make_chains.py` could use multiple cores within a single node (=local), I could use up to 64 cores.  If `make_chains.py` can run on the head node and start submitting single-core jobs (qsub) to our SGE cluster farm, then we could use as many cores as we want.  

Thanks again - I'll ask around and also wait for your test results.  Any more suggestions/hints deeply appreciated!

**Comment by ohdongha** [comment]
> Thanks for your inquiry. I need to do some testing to check whether it's possible in the current make_chains.py setup. I will try to push the make_chains.py job to a single node and check whether nextflow can utilize the node's CPUs within the "local executor" framework. In the best case, Nextflow will "think" that it runs on a machine with 16 CPUs and execute lastz jobs smoothly.
> 
> I will let you know about the results later.

Hello, Bogdan @kirilenkobm,
Have you tried `make_chains.py` in a single node with multiple CPUs and let nextflow to use the node's CPUs as a "local executor" as in your last comment?   

I am still trying to make it work in our SGE cluster with `--executor sge`.  Now nextflow at least tries to submit multiple jobs (before it gave up after realizing `qsub` is not allowed) and fails - progress but not quite there yet.   

In any case, using multiple CPUs within a node as "local executor" would be more desirable since most nodes in our cluster come with 16 ~ 32 CPU cores.  So I would like to hear whether you had any luck.  

Thanks!
Dong-Ha

**Comment by kirilenkobm** [comment]
Hi Dong-Ha @ohdongha,

I appreciate your patience.
I tried this, and everything worked smoothly.

We have a slurm cluster with 20 nodes, ~128 cores on each node.
I wrote a make_chains command that looks as:
`
./make_chains.py hg38 bosTau9 test_inputs/hg38.chr9.fa.gz test_inputs/bosTau9.chr11.fa.gz --project_dir test_paralell_on_node --executor_queuesize 8
`
And saved this into a file that I called "test_job_to_run_on_node."
The command runs on the "local" executed and utilizes at max eight cores.
Then, I pushed this as a single job to be executed on a node using sbatch with 50 Gb of RAM and eight cores limits. This is our in-house solution to push jobs to cluster, but technically it's a wrapper around sbatch:
`
para make test_on_node test_job_to_run_on_node -numCores 8 -memoryMb 60000 -q="short" -maxNumResubmission 0
`

Being pushed on a node, the command is unaware of where it runs, and from its "point of view," it runs on a multi-CPU PC (in simple words). In this context, it utilizes 8 cores on a node as it was a local executor.

Hopefully, it helps; pls let me know if you have any questions.

Best,
Bogdan

**Comment by ohdongha** [comment]
Thanks a lot, Bogdan.  So no need for `--cluster_parameters "cpus=16"` and just limiting the number of cores with `--executor_queuesize 8` would suffice.  Have you checked the walltime vs. CPU time, etc. to see whether multiple cores were used?

I tried the same with our cluster farm, submitting to a single node the following:
`make_chains.py Dme Dsi GCF_000001215.4_Release_6_plus_ISO1_MT_genomic.fna GCF_016746395.2_Prin_Dsim_3.1_genomic.fna --executor_queuesize 16 --project_dir 220722_LastZ_chain_DmeDsi_1stAttempt > 220722_lsz_DmeDsi_1stAttempt.log 2>&1`

And this was the end of the log file just a moment later after the submission.  Do you think the jobs were submitted in parallel?   
...
************ HgStepManager: executing step 'lastz' Fri Jul 22 08:50:27 2022.
doLastzClusterRun ....
testCleanState current step: doLastzClusterRun, /home/ohd3/220622_test_mWGA_tools/220722_LastZ_chain_DmeDsi_1stAttempt/TEMP_run.lastz, /home/ohd3/220622_test_mWGA_tools/220722_LastZ_chain_DmeDsi_1stAttempt/TEMP_run.lastz/lastz.done     previous step: doPartition, /home/ohd3/220622_test_mWGA_tools/220722_LastZ_chain_DmeDsi_1stAttempt/TEMP_run.lastz/partition.done
# chmod a+x /home/ohd3/220622_test_mWGA_tools/220722_LastZ_chain_DmeDsi_1stAttempt/TEMP_run.lastz/doClusterRun.sh
# /home/ohd3/220622_test_mWGA_tools/220722_LastZ_chain_DmeDsi_1stAttempt/TEMP_run.lastz/doClusterRun.sh
+ cd /home/ohd3/220622_test_mWGA_tools/220722_LastZ_chain_DmeDsi_1stAttempt/TEMP_run.lastz
+ gensub2 Dme.lst Dsi.lst gsub jobList
+ parallel_executor.py lastz_DmeDsi jobList -q day --memoryMb 10000 -e local --co None -p batch --eq 16
NOTE: Nextflow is not tested with Java 1.8.0_332 -- It's recommended the use of version 11 up to 18

N E X T F L O W  ~  version 22.04.4
Launching `/home/ohd3/220622_test_mWGA_tools/220722_LastZ_chain_DmeDsi_1stAttempt/TEMP_run.lastz/lastz_DmeDsi/script.nf` [cheeky_edison] DSL1 - revision: 351eeece0b
[-        ] process > execute_jobs [  0%] 0 of 4

executor >  local (4)
[32/b19bd2] process > execute_jobs (1) [  0%] 0 of 4

executor >  local (4)
[32/b19bd2] process > execute_jobs (1) [  0%] 0 of 4
```

Thanks!

**Comment by kirilenkobm** [comment]
That's a good point. I'll need to check whether all 16 cores are utilized.
I'll try to check it today.

**Comment by ohdongha** [comment]
> That's a good point. I'll need to check whether all 16 cores are utilized.
> I'll try to check it today.
>
Inspired by your post, I also started a run comparing larger genomes (human vs. mouse), and it is now obvious that multiple cores are being used!  Perhaps the Drosophila genomes I tried before were too small (I think only four chunks of sequences were created).   I will now just wait for the run to the finish - it seems it will take some time on 32 cores.  

One more question - I am using soft-masked genomes as input and wonder whether this pipeline (or `lastz`) recognizes soft-masked genomes.   If not, will comparing non-hard-masked genomes increase the runtime significantly?  

Thanks again!

**Comment by MichaelHiller** [comment]
Your genomes (both reference and query) should be properly repeat-soft-masked. Without that lastz would take forever. 

We recommend running repeatModeler separately on each genome to get a library and masking it with this lib. The default repeatMasker is often not good enough.

**Comment by kirilenkobm** [comment]
@ohdongha, it can indeed take a while. We run mammalian alignments on 2000 cores, which usually takes ~8 hours. 

About your question about whether nextflow actually runs jobs in parallel being executed directly on a node: I came up with a solution that will allow me to evaluate that, will keep you posted.

**Comment by ohdongha** [comment]
> @ohdongha, it can indeed take a while. We run mammalian alignments on 2000 cores, which usually takes ~8 hours.
> 
> About your question about whether nextflow actually runs jobs in parallel being executed directly on a node: I came up with a solution that will allow me to evaluate that, will keep you posted.

Wow, 2000 cores sound a lot!  

For human-mouse, I think a total of 1690 jobs were created for `doLastzClusterRun`, which were finished in 2d 1h 31m 20s or 687.7 CPU hours (=28.654 CPU days), according to the log file, on a 32-core node.  This was consistent with the SGE job report. 

So on average, 14~15 cores out of 32 were used.   The script indeed runs multi-threaded - this was not obvious for Drosophila pairs because only 4 jobs were created (and most of the 16 or 32 cores assigned were not used - the genomes were too small). 

Yes, please keep me posted about the solution to evaluate the multi-usage - perhaps we can close the issue with it. 

...

There is another issue.   Both Drosophila pair and human-mouse runs failed at `doChainRun` step with the same kind of error.  Since this was not related to multi-threading, I will open another issue and post the log files, etc. there.  

Thanks!

**Comment by ohdongha** [comment]
Now I have successfully run human-vs-mouse and human-vs-zebrafish (and trying to get stats) on a 32-core node.  I can approximately get the information on how many cores were used by dividing the CPU time/wall time.  So I guess I got all I want from raising this issue (for now ;) ).  
Thanks again for your help!

**Comment by drelo** [comment]
Hai all, @ohdongha  since we currently don't have access to a cluster I am trying to use a local Ubuntu 22.04 Desktop with 16 cores (32) threads for this task.  I tried to execute a strategy similar to the one above but  with the toy data given in the _**TOGA**_ github and I couldn't make it work. 

` ../make_lastz_chains/make_chains.py hg mm hg38.2bit mm10.2bit --executor_queuesize 14 --project_dir HERE`

But of course this is just a Desktop with no slurm or other job scheduler... my question is which is the best way to configure SLURM in just a Desktop? [I found this](https://drtailor.medium.com/how-to-setup-slurm-on-ubuntu-20-04-for-single-node-work-scheduling-6cc909574365) but I wonder if this would work? Best regards.
